### PR TITLE
Create very simple state estimator and add depth and orientation esti…

### PR DESCRIPTION
**Pull request by Wesley-Fisher**
Create state estimation node.

Move code to estimate depth and orientation into this node. Use state estimation results in existing controllers.

### CODE REVIEW

File: catkin_ws/src/aquadrone_msgs/msg/SubState.msg
Comment (Waterloo-Aquadrone) [inline catkin_ws/src/aquadrone_msgs/msg/SubState.msg]: Shouldn't we have an orientation variance as well?

Comment (amaarquadri) [inline catkin_ws/src/aquadrone_msgs/msg/SubState.msg:10]: Shouldn't we have orientation variance too?

File: catkin_ws/src/state_estimation/src/simple_state_estimation.py
Comment (amaarquadri) [inline catkin_ws/src/state_estimation/src/simple_state_estimation.py]: What sort of state estimation is this?
From what I can tell, it seems like the position and velocity variances are going to blow up to infinity.
Also, the increment dynamics function should (eventually) take input as to what thrust is being applied by the motors (and torque from buoyancy).

File: catkin_ws/src/state_estimation/src/simple_state_estimation.py
Comment (amaarquadri) [inline catkin_ws/src/state_estimation/src/simple_state_estimation.py]: cb stands for calibration?

File: catkin_ws/src/state_estimation/src/simple_state_estimation.py
Comment (Wesley-Fisher) [reply catkin_ws/src/state_estimation/src/simple_state_estimation.py]: This is a very simple version, that was just meant to get the basics functioning.

The Kalman Filter version that Im working on in another branch (and will probably merge in to replace this) will be using the thruster forces, and likely have damping to prevent velocity from blowing up.

Position will blow up until we have a way to directly measure it.

File: catkin_ws/src/state_estimation/src/simple_state_estimation.py
Comment (Wesley-Fisher) [reply catkin_ws/src/state_estimation/src/simple_state_estimation.py]: callback. I'll look into making it more clear.

File: catkin_ws/src/aquadrone_msgs/msg/SubState.msg
Comment (Wesley-Fisher) [reply catkin_ws/src/aquadrone_msgs/msg/SubState.msg:10]: We should, yeah. I wasnt sure how to do it with a quaternion, but when I was making the Kalman Filter version, I saw some representation of orientation variance. So I'll try to add it in there, and propogate a merge into this.

File: catkin_ws/src/state_estimation/src/ekf.py
Comment (amaarquadri) [inline catkin_ws/src/state_estimation/src/ekf.py:292]: Add comment:
#TODO: Add torque from bouyancy

File: catkin_ws/src/state_estimation/src/ekf.py
Comment (amaarquadri) [inline catkin_ws/src/state_estimation/src/ekf.py:125]: Add comment: # TODO: Add ZED Mini localization data

File: catkin_ws/src/state_estimation/src/ekf.py
Comment (amaarquadri) [inline catkin_ws/src/state_estimation/src/ekf.py:152]: Potential issue: only the most recent motor command is taken into account. If for whatever reason the motor commands change during the course of the time step, then the previous motor command's values will be ignored.
Not sure if this will actually have a measurable effect or not, or if this is actually a scenario that may happen.

File: catkin_ws/src/state_estimation/src/ekf.py
Comment (amaarquadri) [inline catkin_ws/src/state_estimation/src/ekf.py:131]: Hey Wesley, so just to summarize the goal is to run the EKF loop at 20Hz (or another rate based on testing).

Each loop will do a prediction step based on the motor thrusts, gravity, bouyancy, drag, and other forces to update the expected state and its variance. Then a measurement step, where it uses input from the pressure sensor and gyro (and eventually the zed mini and other sensors as well) to refine its expected state and variance. Then there is a step where the new expected state/variance is converted to a SubState message which is then published.

The assumption here is that this entire loop will be able to run within 20Hz every single time (because otherwise the time step being used in calculations would be wrong).

File: catkin_ws/src/state_estimation/src/ekf_indices.py
Comment (amaarquadri) [inline catkin_ws/src/state_estimation/src/ekf_indices.py:3]: Nice addition. Great way to make the code more readable!

File: catkin_ws/src/state_estimation/src/ekf_sensors.py
Comment (amaarquadri) [inline catkin_ws/src/state_estimation/src/ekf_sensors.py:98]: For both the pressure sensor and gyro, the sensor will likely not return variance data. This will likely have to be read off of their data sheets and hardcoded into their respective nodes.

**Review by amaarquadri — CHANGES_REQUESTED** [review]
The EKF looks great. I think it is written very cleanly. I just added a few comments with clarifications and suggestions for comments that should be added into the code, but for the most part this looks great to me.

File: catkin_ws/src/state_estimation/src/ekf.py
Comment (Wesley-Fisher) [reply catkin_ws/src/state_estimation/src/ekf.py:152]: I think there will be some cases where that happens, but it likely won't be a huge issue on these timescales.

A nice thing about the EKF is that we can sort of account for potential issues like this via the uncertainty we assign to the dynamics function.

File: catkin_ws/src/state_estimation/src/ekf.py
Comment (Wesley-Fisher) [reply catkin_ws/src/state_estimation/src/ekf.py:131]: That's a good summary; I might add that as a comment in the code for others.

The timing could be a potential issue in the way you mention. Tracking something like the last time it was run might be better for it.

File: catkin_ws/src/state_estimation/src/ekf_sensors.py
Comment (Wesley-Fisher) [reply catkin_ws/src/state_estimation/src/ekf_sensors.py:98]: Yeah. I'll need to compare the real vs simulated sensors and try to get them to match (the sim depth sensor has variance, and your PR for our depth sensor will be good to help it match).

The IMU node may come with variance data, but I'll need to check that as well.

File: catkin_ws/src/state_estimation/src/ekf.py
Comment (Wesley-Fisher) [reply catkin_ws/src/state_estimation/src/ekf.py:292]: Added

File: catkin_ws/src/state_estimation/src/ekf.py
Comment (Wesley-Fisher) [reply catkin_ws/src/state_estimation/src/ekf.py:125]: Added

File: catkin_ws/src/state_estimation/src/ekf.py
Comment (Wesley-Fisher) [reply catkin_ws/src/state_estimation/src/ekf.py:131]: I've changed how dt is calculated

**Comment by Wesley-Fisher** [comment]
@amaar-quadri Are there any other comments that I've missed, and are still needed before approval/merging?

**Review by amaarquadri — APPROVED** [review]
Looks good to me

**Pull request merged**
